MAINT: distutils: trivial cleanups

xoviat opened the pull request:
From #9975.

charris commented:
Can't you rebase on master and clean that up? Why all the merges and an useless commit?

charris commented:
Note that you can rewrite the history without making a new PR, just clean things up and make a force push.

xoviat commented:
I know perfectly well. But that requires command-line access.

charris commented:
Ah, what platform are you developing on?

xoviat commented:
I'm using the GitHub web editor.

xoviat commented:
It's actually quite entertaining figuring out how to trick GitHub into doing what you want it to. For example, the `[useless commit]` created a new branch.

charris commented:
The GitHub web editor is pretty limited. Do you have other options?

xoviat commented:
There you go.

eric-wieser commented on numpy/core/setup.py:
```diff
@@ -685,13 +685,15 @@ def get_mathlib_info(*args):
-                'extra_compiler_args' : (['/GL-'] if sys.platform == 'win32' else []),
-            })  # empty list required for creating npy_math_internal.h
+                'include_dirs' : [],  # empty list required for creating npy_math_internal.h
+                'extra_compiler_args' : (['/GL-'] if is_msvc else []),
```
This looks like a bugfix, not just cleanup, to me

charris replied:
Is the flag also needed for the Intel compiler? <searches> Looks like Intel has this flag also, so we might want to disable it for that also if Python 3.5 enables it and we use that by default. @cgohlke Thoughts? Clang may be a future concern.

xoviat replied:
Would it be okay to just add them? I am genuinely open to any solution.

xoviat commented:
WTH is wrong with the second commit?

xoviat committed:
MAINT: distutils: trivial cleanups

xoviat committed:
BUG: core: fix setup handling

Check the compiler directly rather than platform
for determining compiler arguments.

xoviat commented:
Can someone restart the failing travis job?

charris commented:
The cleanup looks OK, but I still have concerns about the flag.

xoviat committed:
BUG: distutils: kill npymath LTCG on clang, intel

Disable link-time code generation for intel and clang compilers
if the platform is win32.

xoviat commented:
Flag has been enabled for intel and clang compilers.

charris commented on numpy/core/setup.py:
```diff
@@ -685,13 +685,18 @@ def get_mathlib_info(*args):
+    compiler_type = customized_ccompiler().compiler_type
+    is_msvc = compiler_type == 'msvc'
+    is_msvc = is_msvc or (sys.platform == 'win32' and compiler_type in ('clang', 'intel'))
+
```
I still think platform may be the way to go, but not `sys.platform`, rather the module.
```
import platform
platform.system() == 'Windows'
```

xoviat committed:
BUG: core: check platform.system for npymath

xoviat commented:
That should be OK. The `is_msvc` variable is somewhat misleading but it's defined right above and comes with a comment.

charris commented:
OK, let's give this a shot and keep an eye out for possible problems. This isn't something easy for us to test.

Thanks @xoviat .

0017031 commented:
But this "is_msvc" flag bugged me  when compiling numpy under mingw-w64-x86_64-python3 which is compiled by `GCC 8.2.1 20181214 64 bit (AMD64)`.

It feels strange deciding a **compiler** flag by `platform.system()`.
Would `platform.python_compiler()` be better?

And, by the way, according to msdn, /GL is default off.

> Whole program optimization is off by default and must be explicitly enabled

rgommers commented:
@0017031 we'd like to support mingw-w64 compilers as well as possible. If you ran into a problem, can you please open a new issue or PR? I'm not sure if "bugged" means that a fix is needed or it just looked odd to you.

> It feels strange deciding a compiler flag by platform.system().
Would platform.python_compiler() be better?

Note that all of distutils feels strange and is hard to test/support. So whatever works. If your proposed change solves an issue for mingw and still works for MSVC then yes let's change it. If it's just cosmetic then no.

0017031 commented:
Sorry for the unclear description.

Today I tried to install numpy with mingw-w64-x86_64-python3 , when it calling gcc to compile, the error-log reported "unknown compiler flag '/GL-' ", so I looked into the source and traced disscussion here.

Here in the setup.py , "is_msvc" acts as a flag controlling whether the 'extra_compiler_args' should be '`/GL-`' or empty. ` /GL-` is a msvc compiler option. Gcc can't recognize it.

From commit [BUG: core: check platform.system for npymath](https://github.com/numpy/numpy/commit/2a26477c67bf319428ddcb78a5b6a8777754b845)
  ```
  is_msvc = platform.system() == 'Windows'
    config.add_installed_library('npymath',
            sources=npymath_sources + [get_mathlib_info],
            install_dir='lib',
            build_info={
                'include_dirs' : [],  # empty list required for creating npy_math_internal.h
                'extra_compiler_args' : (['/GL-'] if is_msvc else []),
            })
```

In my opnion, platform.system() won't be a good choice deciding which compiler is in use.
It reports **system** information.
If I wrote the code here, I'd choose `platform.python_compiler()`.

For the python installation I have now on hand:
mingw-w64-x86_64-python3, platform.python_compiler()  reports GCC 8.2.1 20181214 64 bit (AMD64)
python-3.7.3 [windows] reports MSC v.1916 32 bit (Intel)
python2.7.14 [windows] reports MSC v.1500 32 bit (Intel)

rgommers commented:
Thanks @0017031, clear now. How about something like:
```
is_msvc = platform.platform.startswith('Windows') and platform.python_compiler.startswith('MS'):
```

rgommers commented:
Or perhaps we already have a function to identify MSVC, I have some memory of that. That must be useful in more places.
